0.preprocessing_data/scripts/2.z_slice_instensity_normalization.py

```diff
@@ -0,0 +1,379 @@
+    # loop through each slice and get the nth percentile pixel value
+    for slice in range(image.shape[0]):
+        original_image_nth_pixel_dict["slice"].append(slice)
+        original_image_nth_pixel_dict[f"{percentile}_percentile_pix"].append(
+            pd.Series(image[slice, :, :].flatten()).quantile(percentile / 100)
+        )
+    original_image_nth_pixel_df = pd.DataFrame(original_image_nth_pixel_dict)
+
+    # find the slice with the highest 99th percentile pixel value
+    max_percentile_slice = original_image_nth_pixel_df.loc[
+        original_image_nth_pixel_df[f"{percentile}_percentile_pix"].idxmax()
+    ]
+    max_percentile_slice = image[max_percentile_slice["slice"].astype(int), :, :]
+
+    # perform histogram matching on all slices to the slice with the highest 99th percentile pixel value
+    # new image to store the matched histogram slices
+    new_image = np.zeros(image.shape)
+    for slice in range(image.shape[0]):
+        new_image[slice, :, :] = match_histograms(
+            image[slice, :, :], max_percentile_slice, channel_axis=None
+        )
+    new_image = new_image.astype(np.uint16)
+
+    # get the nth percentile of the normalized image for each slice
+    original_image_nth_pixel_normalized_dict = {
+        "slice": [],
+        f"{percentile}_percentile_pix": [],
+    }
+    for slice in range(new_image.shape[0]):
+        original_image_nth_pixel_normalized_dict["slice"].append(slice)
+        original_image_nth_pixel_normalized_dict[f"{percentile}_percentile_pix"].append(
+            pd.Series(new_image[slice, :, :].flatten()).quantile(percentile / 100)
+        )
+    original_image_nth_pixel_normalized_df = pd.DataFrame(
+        original_image_nth_pixel_normalized_dict
+    )
+
+    if return_df:
+        return (
+            new_image,
+            original_image_nth_pixel_df,
+            original_image_nth_pixel_normalized_df,
+        )
+    else:
+        return new_image
+
+
+def plot_max_ortho_slices(
+    image: np.ndarray,
+    anisotropy: Tuple[int, None] = None,
+    show_plot: bool = True,
+    plot_output_path: Tuple[pathlib.Path, None] = None,
+) -> None:
+    """
+    A function to plot the maximum intensity projection of an image along the x, y, and z axes.
+
+    Parameters
+    ----------
+    image : np.ndarray
+        A 3D numpy array representing an image.
+        The dimensions of the array should be (z, y, x).
+    anisotropy : Tuple[int, None], optional
+        A tuple of two integers representing the anisotropy of the image in the x and y directions.
+        The default is None.
+    show_plot : bool, optional
+        A boolean indicating whether to show the plot or not.
+        The default is True.
+    plot_output_path : pathlib.Path, optional
+        A pathlib.Path object representing the path to save the plot.
+        The default is None.
+
+    Returns
+    -------
+    None
```

Comment:
Even though the function doesn’t return anything, adding an explanation of the generated plot could significantly enhance readability. Providing context, such as what the plot represents, the meaning of the axes, and any key takeaways, would help readers better understand the purpose and insights the visualization offers. For example, you could briefly describe what each axis represents, what the plot conveys about the data, and how it might be interpreted in the broader analysis.

This would make it easier for readers to grasp the value of the plot at a glance, especially if they’re unfamiliar with the data or the context of the visualization.